pkg/pillar/zedcloud/authen.go

```diff
@@ -384,68 +385,92 @@ func VerifySigningCertChain(ctx *ZedCloudContext, content []byte) ([]byte, error
 	err := proto.Unmarshal(content, sm)
 	if err != nil {
 		log.Errorf("SaveCertsToFile: Unmarshal err %v\n", err)
-		return nil, err
+		return []byte{}, err
 	}
 
-	var block *pem.Block
-	var leafcert *x509.Certificate
-	var certByte []byte
-	var keyCnt int
+	// prepare intermediate certs and validate the payload
+	var sigCertBytes []byte
+	var keyCnt, signKeyCnt int
 	interm := x509.NewCertPool()
 	for _, cert := range sm.GetCerts() {
 		keyCnt++
-		if cert.Type == zcert.ZCertType_CERT_TYPE_CONTROLLER_INTERMEDIATE {
+		switch cert.Type {
+		case zcert.ZCertType_CERT_TYPE_CONTROLLER_INTERMEDIATE:
 			ok := interm.AppendCertsFromPEM(cert.GetCert())
 			if !ok {
-				err := fmt.Errorf("VerifySigningCertChain: fail to append intermediate certs")
-				return nil, err
-			}
-			log.Debugf("VerifySigningCertChain: get intermediate cert %d, len %d\n",
-				keyCnt, len(cert.GetCert()))
-		} else if cert.Type == zcert.ZCertType_CERT_TYPE_CONTROLLER_SIGNING {
-			if len(certByte) > 0 {
-				err := fmt.Errorf("VerifySigningCertChain: received more than one leaf cert")
-				return nil, err
-			}
-			certByte = cert.GetCert()
-			block, _ = pem.Decode(certByte)
-			if block == nil {
-				err := fmt.Errorf("VerifySigningCertChain: can not get cert block")
-				return nil, err
+				errStr := fmt.Sprintf("intermediate cert append fail")
+				log.Errorf("VerifySigningCertChain: " + errStr)
+				return []byte{}, errors.New(errStr)
 			}
-			leafcert, err = x509.ParseCertificate(block.Bytes)
-			if err != nil {
-				log.Errorf("VerifySigningCertChain: fail to parse certificate %v\n", err)
-				return nil, err
+
+		case zcert.ZCertType_CERT_TYPE_CONTROLLER_SIGNING:
+			signKeyCnt++
+			if signKeyCnt > 1 {
+				errStr := fmt.Sprintf("received more than one signing cert")
+				log.Errorf("VerifySigningCertChain: " + errStr)
+				return []byte{}, errors.New(errStr)
 			}
-			log.Debugf("VerifySigningCertChain: get signing cert %d, len %d\n",
-				keyCnt, len(cert.GetCert()))
-		} else if cert.Type == zcert.ZCertType_CERT_TYPE_CONTROLLER_ECDH_EXCHANGE {
-			// skip for the CONTROLLER_ECDH_EXECHANGE cert type here, not for signing purpose
-			log.Debugf("VerifySigningCertChain: ECDH_EXCHANGE type, ignore cert %d, len %d\n",
-				keyCnt, len(cert.GetCert()))
-		} else {
-			err := fmt.Errorf("VerifySigningCertChain: received cert with unknown type %d", cert.Type)
-			return nil, err
+			sigCertBytes = cert.GetCert()
+
+		case zcert.ZCertType_CERT_TYPE_CONTROLLER_ECDH_EXCHANGE:
+			// nothing to do
```

Comment:
Can we handle multiple of these ECDH certificates?

Reply:
we can, for future considerations